File: web/src/lib/miniSiteVideo.ts
import type { MiniSiteTemplate } from "@/types/miniSite";

export type MiniSiteVideoProvider = "youtube" | "vimeo";

export type MiniSiteVideoMedia = {
  kind: "video";
  url: string;
  provider: MiniSiteVideoProvider;
  embedUrl: string;
  title: string;
};

export type MiniSiteVideoMediaWire = {
  kind: "video";
  url: string;
  provider: MiniSiteVideoProvider;
  embed_url: string;
  title?: string;
};

export const MINI_SITE_VIDEO_INVALID_URL_MESSAGE = "Use a YouTube or Vimeo link.";

const ALLOWED_EMBED_HOSTS = new Set(["www.youtube.com", "youtube.com", "player.vimeo.com"]);

export type MiniSiteTemplateVideos = Partial<Record<string, MiniSiteVideoMedia>>;

export function parseMiniSiteVideoUrl(input: string): Omit<MiniSiteVideoMedia, "kind" | "title"> | null {
  const raw = input.trim();
  if (!raw) {
    return null;
  }

  let parsed: URL;
  try {
    parsed = new URL(raw);
  } catch {
    return null;
  }

  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    return null;
  }

  const host = parsed.hostname.toLowerCase();
  const path = parsed.pathname;

  if (host === "www.youtube.com" || host === "youtube.com") {
    if (path.startsWith("/embed/")) {
      const videoId = path.slice("/embed/".length).split("/")[0]?.split("?")[0];
      if (videoId) {
        return youtubeResult(raw, videoId);
      }
    }
    if (path === "/watch" || path.startsWith("/watch/")) {
      const videoId = parsed.searchParams.get("v")?.split("&")[0];
      if (videoId) {
        return youtubeResult(raw, videoId);
      }
    }
  }

  if (host === "youtu.be") {
    const videoId = path.replace(/^\//, "").split("/")[0]?.split("?")[0];
    if (videoId) {
      return youtubeResult(raw, videoId);
    }
  }

  if (host === "www.vimeo.com" || host === "vimeo.com") {
    const match = path.match(/^\/(\d+)/);
    if (match?.[1]) {
      return {
        provider: "vimeo",
        url: raw,
        embedUrl: `https://player.vimeo.com/video/${match[1]}`,
      };
    }
  }

  return null;
}

function youtubeResult(url: string, videoId: string): Omit<MiniSiteVideoMedia, "kind" | "title"> {
  return {
    provider: "youtube",
    url,
    embedUrl: `https://www.youtube.com/embed/${videoId}`,
  };
}

export function isAllowedMiniSiteVideoEmbedUrl(embedUrl: string): boolean {
  try {
    const parsed = new URL(embedUrl);
    if (parsed.protocol !== "https:") {
      return false;
    }
    const host = parsed.hostname.toLowerCase();
    if (!ALLOWED_EMBED_HOSTS.has(host)) {
      return false;
    }
    if (host === "www.youtube.com" || host === "youtube.com") {
      return parsed.pathname.startsWith("/embed/");
    }
    if (host === "player.vimeo.com") {
      return parsed.pathname.startsWith("/video/");
    }
    return false;
  } catch {
    return false;
  }
}

export function normalizeMiniSiteVideoMedia(value: unknown): MiniSiteVideoMedia | null {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }

  const source = value as Record<string, unknown>;
  if (source.kind !== undefined && source.kind !== "video") {
    return null;
  }

  const urlRaw = source.url;
  if (typeof urlRaw !== "string") {
    return null;
  }
  const url = urlRaw.trim();
  if (!url) {
    return null;
  }

  const parsed = parseMiniSiteVideoUrl(url);
  if (parsed) {
    return {
      kind: "video",
      url: parsed.url,
      provider: parsed.provider,
      embedUrl: parsed.embedUrl,
      title: typeof source.title === "string" ? source.title.trim() : "",
    };
  }

  const providerRaw = source.provider;
  const embedUrlRaw = source.embedUrl ?? source.embed_url;
  if (
    (providerRaw === "youtube" || providerRaw === "vimeo") &&
    typeof embedUrlRaw === "string" &&
    isAllowedMiniSiteVideoEmbedUrl(embedUrlRaw.trim())
  ) {
    return {
      kind: "video",
      url,
      provider: providerRaw,
      embedUrl: embedUrlRaw.trim(),
      title: typeof source.title === "string" ? source.title.trim() : "",
    };
  }

  return null;
}

export function mapMiniSiteVideoMediaToWire(media: MiniSiteVideoMedia): MiniSiteVideoMediaWire {
  return {
    kind: "video",
    url: media.url,
    provider: media.provider,
    embed_url: media.embedUrl,
    title: media.title || undefined,
  };
}

export function buildMiniSiteVideoMediaFromUrl(input: string, title = ""): MiniSiteVideoMedia | null {
  const parsed = parseMiniSiteVideoUrl(input);
  if (!parsed) {
    return null;
  }
  return {
    kind: "video",
    ...parsed,
    title: title.trim(),
  };
}

export function getTemplateVideoSlots(
  templateMedia: Partial<Record<MiniSiteTemplate, Record<string, unknown>>>,
  template: MiniSiteTemplate,
): MiniSiteTemplateVideos {
  const bucket = templateMedia[template];
  if (!bucket || typeof bucket !== "object") {
    return {};
  }

  const result: MiniSiteTemplateVideos = {};
  for (const [slotKey, slotValue] of Object.entries(bucket)) {
    const media = normalizeMiniSiteVideoMedia(slotValue);
    if (media) {
      result[slotKey] = media;
    }
  }
  return result;
}

File: api/app/utils/mini_site_video.py
"""Safe YouTube/Vimeo URL parsing for mini-site template video slots."""

from __future__ import annotations

import re
from typing import Any, Literal
from urllib.parse import parse_qs, urlparse

MiniSiteVideoProvider = Literal["youtube", "vimeo"]

_ALLOWED_EMBED_HOSTS = frozenset({"www.youtube.com", "youtube.com", "player.vimeo.com"})


def _sanitize_text(value: object) -> str:
    if not isinstance(value, str):
        return ""
    return value.replace("<", "").replace(">", "").strip()


def parse_mini_site_video_url(input_value: str) -> dict[str, str] | None:
    raw = _sanitize_text(input_value)
    if not raw:
        return None

    try:
        parsed = urlparse(raw)
    except ValueError:
        return None

    if parsed.scheme not in {"http", "https"}:
        return None

    host = (parsed.hostname or "").lower()
    path = parsed.path or ""

    if host in {"www.youtube.com", "youtube.com"}:
        if path.startswith("/embed/"):
            video_id = path.removeprefix("/embed/").split("/")[0].split("?")[0]
            if video_id:
                return _video_result("youtube", raw, f"https://www.youtube.com/embed/{video_id}")
        if path == "/watch" or path.startswith("/watch/"):
            query = parse_qs(parsed.query)
            video_ids = query.get("v")
            if video_ids and video_ids[0]:
                video_id = video_ids[0].split("&")[0]
                if video_id:
                    return _video_result("youtube", raw, f"https://www.youtube.com/embed/{video_id}")

    if host == "youtu.be":
        video_id = path.strip("/").split("/")[0].split("?")[0]
        if video_id:
            return _video_result("youtube", raw, f"https://www.youtube.com/embed/{video_id}")

    if host in {"www.vimeo.com", "vimeo.com"}:
        match = re.match(r"^/(\d+)", path)
        if match:
            video_id = match.group(1)
            return _video_result("vimeo", raw, f"https://player.vimeo.com/video/{video_id}")

    return None


def _video_result(provider: MiniSiteVideoProvider, url: str, embed_url: str) -> dict[str, str]:
    return {
        "provider": provider,
        "url": url,
        "embed_url": embed_url,
    }


def is_allowed_mini_site_video_embed_url(embed_url: str) -> bool:
    try:
        parsed = urlparse(embed_url)
    except ValueError:
        return False
    if parsed.scheme != "https":
        return False
    host = (parsed.hostname or "").lower()
    if host not in _ALLOWED_EMBED_HOSTS:
        return False
    if host in {"www.youtube.com", "youtube.com"}:
        return parsed.path.startswith("/embed/")
    if host == "player.vimeo.com":
        return parsed.path.startswith("/video/")
    return False


def normalize_video_media_value(value: object) -> dict[str, Any] | None:
    if not isinstance(value, dict):
        return None

    kind = value.get("kind")
    if kind is not None and kind != "video":
        return None

    url_raw = value.get("url")
    if not isinstance(url_raw, str):
        return None
    url = _sanitize_text(url_raw)
    if not url:
        return None

    parsed = parse_mini_site_video_url(url)
    if parsed is None:
        embed_raw = value.get("embed_url", value.get("embedUrl"))
        provider_raw = value.get("provider")
        if (
            isinstance(embed_raw, str)
            and isinstance(provider_raw, str)
            and provider_raw in {"youtube", "vimeo"}
            and is_allowed_mini_site_video_embed_url(_sanitize_text(embed_raw))
        ):
            return {
                "kind": "video",
                "url": url,
                "provider": provider_raw,
                "embed_url": _sanitize_text(embed_raw),
                "title": _sanitize_text(value.get("title")) or "",
            }
        return None

    return {
        "kind": "video",
        "url": parsed["url"],
        "provider": parsed["provider"],
        "embed_url": parsed["embed_url"],
        "title": _sanitize_text(value.get("title")) or "",
    }
